src/metabase/models/card.clj

```diff
@@ -654,12 +654,88 @@
+                 expressions))
+
+(defn- ensure-clause-idents-joins [joins stage-number {:keys [entity_id] :as _ctx}]
+  (mapv #(assoc % :ident (derive-ident "join" entity_id stage-number (:alias %)))
+        joins))
+
+(defn- ensure-clause-idents-inner [inner-query ctx]
+  (let [{:keys [query stage-number]} (if-let [source-query (:source-query inner-query)]
+                                       (-> (ensure-clause-idents-inner source-query ctx)
+                                           (update :query #(assoc inner-query :source-query %)))
+                                       {:query inner-query
+                                        :stage-number 0})]
+    {:stage-number (inc stage-number)
+     :query (cond-> query
+              (:aggregation query) (update :aggregation-idents
+                                           ensure-clause-idents-list (:aggregation query) "aggregation" stage-number ctx)
+              (:expressions query) (update :expression-idents
+                                           ensure-clause-idents-expressions (:expressions query) stage-number ctx)
+              (:breakout query)    (update :breakout-idents
+                                           ensure-clause-idents-list (:breakout query) "breakout" stage-number ctx)
+              (:joins query)       (update :joins       ensure-clause-idents-joins stage-number ctx))}))
+
+(defn- ensure-clause-idents-outer [{:keys [query type] :as outer-query} ctx]
+  (if-let [{inner-query :query} (when (and (= type :query) query)
+                                  (ensure-clause-idents-inner query ctx))]
+    (assoc outer-query :query inner-query)
+    outer-query))
+
+(defn- ensure-clause-idents
+  ([card]
+   (ensure-clause-idents card nil))
+  ([card forced-entity-id]
+   (let [hashed-eid (atom false)
+         entity-id  (or forced-entity-id
+                        (:entity_id card)
+                        (delay
+                          (when-not (every? #(contains? card %) [:collection_id :created_at :name])
+                            (throw (ex-info "Best-effort backfill of :entity_id on :model/Card failed"
+                                            {::best-effort-backfill true})))
+                          (reset! hashed-eid true)
+                          (serdes/backfill-entity-id card)))
+         ;; Calculate the new query. If we need to backfill its `:entity_id` but fails, it will throw a specific
+         ;; exception that gets caught here, making query' be nil. In that case the query is not changed.
+         query'    (when (:dataset_query card)
+                     (try
+                       (ensure-clause-idents-outer (:dataset_query card) {:entity_id entity-id})
+                       (catch clojure.lang.ExceptionInfo e
+                         (if (::best-effort-backfill (ex-data e))
+                           (log/warnf "Best-effort backfill of :entity_id failed for Card %d" (:id card))
```

Comment:
```suggestion
                           (do (log/warnf "Best-effort backfill of :entity_id failed for Card %d" (:id card))
                               nil)
```
In situations where the return value influences the control flow, I prefer to be explicit, so as not to rely on logging functions returning `nil` (or `false`) and also that the reader doesn't have to go and check that.

Reply:
I've rearranged this to make the re-`throw` an early exit; now there's an explicit `nil` return.